server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardChangesObserver.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.routing.RoutingChangesObserver;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+
+public class ShardChangesObserver implements RoutingChangesObserver {
+
+    private static final Logger logger = LogManager.getLogger(ShardChangesObserver.class);
```

Comment:
I would like to start a discussion here re the log level we should be logging the shard changes. I see the following possibilities:
* log all at info (currently implemented)
* log all at debug so you need to opt in to get those details (even though they might be very helpful in sum investigations)
* log all movements as log as we have few shards (may be less then 1000?)
* have include/exclude pattern (similarly to https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#transport-tracer)

Reply:
I don't think we can reasonably log this volume at `INFO`. We may want to write it to a system index or data stream. For now I vote for `DEBUG`.